hack/tools/e2e-log-sync/main.go

```diff
@@ -0,0 +1,327 @@
+	}
+
+	if err := importLogs(); err != nil {
+		fmt.Printf("Failed to import logs: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+func validateAndCompleteFlags() error {
+	if *logBucket == "" {
+		fmt.Printf("--log-bucket must not be empty\n")
+	}
+	if *logFolder == "" && *logProwJobURL == "" {
+		fmt.Printf("either --log-folder or --log-prow-job-url must not be empty\n")
+	}
+	if *logFolder != "" && *logProwJobURL != "" {
+		fmt.Printf("only one of --log-folder or --log-prow-job-url can be set\n")
+	}
+
+	if *logProwJobURL != "" {
+		var err error
+		*logBucket, *logFolder, err = calculateLogLocation(*logProwJobURL)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func calculateLogLocation(prowJobURL string) (string, string, error) {
+	u, err := url.Parse(prowJobURL)
+	if err != nil {
+		return "", "", errors.Wrapf(err, "failed to parse ProwJob URL %q", prowJobURL)
+	}
+
+	// Path has e.g. format:
+	// /view/gs/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api/6189/pull-cluster-api-build-main/1496233282759561216
+	p := strings.TrimPrefix(u.Path, "/")
+	pathSegments := strings.Split(p, "/")
+
+	bucket := pathSegments[2]
+	folder := path.Join(pathSegments[3:]...)
+
+	return bucket, folder, nil
+}
+
+func importLogs() error {
+	ctx := context.Background()
+
+	// Create GCS client.
+	client, err := storage.NewClient(ctx, option.WithoutAuthentication())
+	if err != nil {
+		return fmt.Errorf("failed to create GCS client: %v", err)
+	}
+	defer client.Close()
+
+	// Get logs from GCS
+	logs, err := getLogsFromGCS(ctx, client, *logBucket, *logFolder)
+	if err != nil {
+		return errors.Wrapf(err, "failed to get logs from GCS")
+	}
+
+	for logFile, ld := range logs {
+		// Prepare logs for Loki.
+		streams, err := prepareDataForLoki(ld)
+		if err != nil {
+			return errors.Wrapf(err, "failed to prepare data for Loki")
+		}
+
+		// Push logs to Loki.
+		if err := pushLogsToLoki(ctx, logFile, streams); err != nil {
+			return errors.Wrapf(err, "failed to push data to Loki")
+		}
+	}
+
+	return nil
+}
+
+// LogData represents the log file and metadata.
+type LogData struct {
+	logs     []byte
+	metadata []byte
+	fileName string
+}
+
+// getLogsFromGCS gets logs from GCS.
+func getLogsFromGCS(ctx context.Context, client *storage.Client, bucket, folder string) (map[string]LogData, error) {
+	klog.Infof("Getting logs from gs://%s/%s", bucket, folder)
+
+	// Create iterator to iterate over GCS files in folder.
+	query := &storage.Query{
+		Prefix:    folder,
+		Delimiter: "",
+	}
+	it := client.Bucket(bucket).Objects(ctx, query)
+
+	// Iterate over files and collect LogData.
+	data := map[string]LogData{}
+	for {
+		attrs, err := it.Next()
+		if err != nil {
+			// Break if there are no other files.
+			if err == iterator.Done {
+				break
+			}
+			return nil, errors.Wrapf(err, "failed to get next file")
+		}
+		fileName := attrs.Name
+
+		// Continue if we found a file other than manager.log.
+		if !strings.HasSuffix(fileName, "manager.log") {
```

Comment:
Should we grab *.log ?

Reply:
I made that part configurable. Default is still manager.log because otherwise we also pull in logs like clusterctl-init.log.

But with the right regex we can now also import other logs (can be relevant in cases like crashd where log files have the name <container-name>.log)